_source/_posts/2022-01-26-webassembly-on-kubernetes-with-rust.md

```diff
@@ -0,0 +1,365 @@
+          value: info
+        - name: RUST_BACKTRACE
+          value: "1"
+  tolerations:
+    - key: "node.kubernetes.io/network-unavailable"
+      operator: "Exists"
+      effect: "NoSchedule"
+    - key: "kubernetes.io/arch"
+      operator: "Equal"
+      value: "wasm32-wasi"
+      effect: "NoExecute"
+    - key: "kubernetes.io/arch"
+      operator: "Equal"
+      value: "wasm32-wasi"
+      effect: "NoSchedule"
+```
+
+**Note**: Remember to replace `<your GitHub user>` with your own GitHub username.
+
+The `annotations` and `tolerations` are significant. The annotations are used to allow external network calls from the krustlet, and the tolerations limit the pod to only schedule/run on Wasm nodes. We are also passing some environment variables that the application will use.
+
+Now apply the manifest using the below command and check the pod status.
+
+```bash
+kubectl apply -f k8s.yaml
+
+kubectl get pods -o wide
+# Output
+NAME                READY   STATUS    RESTARTS   AGE     IP       NODE       NOMINATED NODE   READINESS GATES
+rust-wasi-example   1/1     Running   0          6m50s   <none>   krustlet   <none>           <none>
+```
+
+You should see the Wasm workload running successfully on the Krustlet node. Let's check the logs.
+
+```bash
+kubectl logs rust-wasi-example
+# Output
+[2022-01-16T11:42:20Z INFO  rust_wasm] Cat Fact: Polydactyl cats (a cat with 1-2 extra toes on their paws) have this as a result of a genetic mutation. These cats are also referred to as 'Hemingway cats' because writer Ernest Hemingway reportedly owned dozens of them at his home in Key West, Florida.
+[2022-01-16T11:43:21Z INFO  rust_wasm] Cat Fact: The way you treat kittens in the early stages of it's life will render it's personality traits later in life.
+```
+
+Awesome. You have successfully created a Wasm workload using Rust and deployed it to a Kubernetes cluster without using containers.
+If you'd like to take a look at this solution in full, check out the [GitHub repo](https://github.com/oktadev/okta-rust-webassembly-k8s-example).
+
+# So, are we ready to replace containers with WebAssembly?
+
+WebAssembly on Kubernetes is still not production-ready as a lot of the supporting ecosystem is still experimental, and WASI itself is still maturing. [Networking is still not stable](https://radu-matei.com/blog/towards-sockets-networking-wasi/) and the library ecosystem is only just coming along. Krustlet is also still in beta, and there is no straightforward way to run networking workloads, especially servers on it. [WasmEdge](https://wasmedge.org/) is another more mature solution for networking workloads, but it's much more involved to setup and run than Krustlet on Kubernetes. [WasmCloud](https://wasmcloud.dev/) is another project to keep an eye on. So, for the time being, Krustlet is suitable for running workloads for Jobs and use cases involving cluster monitoring and so on, and these are areas you could use the extra performance anyway.
```

Comment:
```suggestion
WebAssembly on Kubernetes is not yet production-ready as a lot of the supporting ecosystem is still experimental, and WASI itself is still maturing. [Networking is not yet stable](https://radu-matei.com/blog/towards-sockets-networking-wasi/) and the library ecosystem is only just coming along. Krustlet is also still in beta, and there is no straightforward way to run networking workloads, especially servers on it. [WasmEdge](https://wasmedge.org/) is a more mature alternative solution for networking workloads, but it's much more involved to set up and run than Krustlet on Kubernetes. [WasmCloud](https://wasmcloud.dev/) is another project to keep an eye on. So, for the time being, Krustlet is suitable for running workloads for jobs and use cases involving cluster monitoring and so on. These are areas where you could use the extra performance anyway.
```
avoiding the repetition of "still" and "another" with some synonyms
closing w/a shorter sentence